kafka-rest/src/main/java/io/confluent/kafkarest/backends/kafka/KafkaModule.java

```diff
@@ -33,35 +37,48 @@
  *
  * <p>Right now this module does little but delegate to {@link KafkaRestContext}, since access to
  * Kafka is currently being configured there. It's the author's intention to move such logic here,
- * and eliminate {@code KafkaRestContext}, once dependence injection is properly used elsewhere.</p>
+ * and eliminate {@code KafkaRestContext}, once dependence injection is properly used elsewhere.
  */
 public final class KafkaModule extends AbstractBinder {
 
+  public static final String KAFKA_REST_CONTEXT_PROPERTY_NAME = "io.confluent.kafkarest.context";
+
   @Override
   protected void configure() {
-    bindFactory(KafkaRestContextFactory.class)
+    bindFactory(KafkaRestContextFactory.class, Singleton.class)
         .to(KafkaRestContext.class)
         .in(RequestScoped.class);
 
-    bindFactory(AdminFactory.class)
-        .to(Admin.class)
-        .in(RequestScoped.class);
+    bindFactory(AdminFactory.class).to(Admin.class).in(RequestScoped.class);
 
     bindFactory(ProducerFactory.class)
-        .to(new TypeLiteral<Producer<byte[], byte[]>>() { })
+        .to(new TypeLiteral<Producer<byte[], byte[]>>() {})
         .in(RequestScoped.class);
   }
 
   private static final class KafkaRestContextFactory implements Factory<KafkaRestContext> {
 
+    private final Provider<ContainerRequestContext> requestContext;
+    private final KafkaRestContext defaultContext;
+
+    @Inject
+    private KafkaRestContextFactory(
+        @Context Provider<ContainerRequestContext> requestContext, KafkaRestConfig config) {
+      this.requestContext = requireNonNull(requestContext);
+      this.defaultContext = new DefaultKafkaRestContext(config);
+    }
+
     @Override
     public KafkaRestContext provide() {
-      return KafkaRestContextProvider.getCurrentContext();
+      KafkaRestContext context =
+          (KafkaRestContext) requestContext.get().getProperty(KAFKA_REST_CONTEXT_PROPERTY_NAME);
+      return context != null ? context : defaultContext;
```

Comment:
So just for my own understanding, would we be hitting always the same factory singleton instance (and therefore the same provider and default context instance), but `requestContext.get()` will be returning request-scoped instance (different `ContainerRequestContext` for different requests)?